pkg/volume/util/atomic_writer.go

```diff
@@ -0,0 +1,305 @@
+// 8.  The previous timestamped directory is removed, if it exists
+// 9.  The sentinel file modification and access times are updated (file is created if it does not
+//     already exist)
+func (w *AtomicWriter) Write(payload map[string][]byte) error {
+	shouldWritePayload, err := w.shouldWritePayload(payload)
+	if err != nil {
+		glog.Errorf("%v: unable to determine whether payload should be written to disk: %v", w.LogContext, err)
+		return err
+	}
+
+	if !shouldWritePayload {
+		glog.V(5).Infof("%v: No update required for target directory %v", w.LogContext, w.TargetDir)
+		return nil
+	}
+
+	tsDir, err := w.newTimestampDir()
+	if err != nil {
+		glog.V(5).Infof("%v: unable to create new timestamped directory: %v", w.LogContext, err)
+		return err
+	}
+
+	err = w.writePayloadToDir(payload, tsDir)
+	if err != nil {
+		glog.Errorf("%v: unable to write payload to dir %v: %v", w.LogContext, tsDir, err)
+		return err
+	}
+
+	err = w.createPayloadSymlinksInTargetDir(payload)
+	if err != nil {
+		glog.Errorf("%v: unable to create payload symlinks in %v: %v", w.LogContext, w.TargetDir, err)
+		return err
+	}
+
+	_, tsDirName := filepath.Split(tsDir)
+	currentStateDirPath := path.Join(w.TargetDir, w.currentStateDirName())
+	oldTsDir, err := os.Readlink(currentStateDirPath)
+	if err != nil && !os.IsNotExist(err) {
+		glog.Errorf("%v: unable to read link for current state directory: %v", w.LogContext, err)
+		return err
+	}
+
+	updateStateDirPath := path.Join(w.TargetDir, w.updateStateDirName())
+	err = os.Symlink(tsDirName, updateStateDirPath)
+	if err != nil {
+		os.RemoveAll(tsDir)
+		glog.Errorf("%v: unable to create symolic link: %s", w.LogContext, err.Error())
+		return err
+	}
+
+	err = os.Rename(updateStateDirPath, currentStateDirPath)
+	if err != nil {
+		os.Remove(updateStateDirPath)
+		os.RemoveAll(tsDir)
+		glog.Errorf("%v: unable to rename symbolic link %v: %v", w.LogContext, updateStateDirPath, err)
+		return err
+	}
+	if len(oldTsDir) > 0 {
+		err = os.RemoveAll(path.Join(w.TargetDir, oldTsDir))
+		if err != nil {
+			glog.Errorf("%v: unable to remove directory %v: %v", w.LogContext, oldTsDir, err)
+		}
+	}
+
+	// TODO: determine error handling around sentinel file
+	return w.touchSentinelFile()
+}
+
+// shouldWritePayload returns whether the payload must be written to disk
+func (w *AtomicWriter) shouldWritePayload(payload map[string][]byte) (bool, error) {
+	for projectionPath, content := range payload {
+		shouldWrite, err := w.shouldWriteFile(path.Join(w.TargetDir, path.Clean(projectionPath)), content)
+		if err != nil {
+			return false, err
+		}
+
+		if shouldWrite {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+// shouldWriteFile returns whether a new version of a file should be written to disk.
+func (w *AtomicWriter) shouldWriteFile(path string, content []byte) (bool, error) {
+	_, err := os.Lstat(path)
+	if err != nil && os.IsNotExist(err) {
+		return true, nil
```

Comment:
This one can be safely written as just the call to `os.IsNotExist`, I will change